libnavigation-core/src/main/java/com/mapbox/navigation/core/routeoptions/MapboxRouteOptionsProvider.kt

```diff
@@ -1,21 +1,47 @@
-package com.mapbox.navigation.core.directions.session
+package com.mapbox.navigation.core.routeoptions
 
 import android.location.Location
 import com.mapbox.api.directions.v5.models.RouteOptions
+import com.mapbox.base.common.logger.Logger
+import com.mapbox.base.common.logger.model.Message
+import com.mapbox.base.common.logger.model.Tag
 import com.mapbox.geojson.Point
 import com.mapbox.navigation.base.trip.model.RouteProgress
-import com.mapbox.navigation.core.trip.session.TripSession
 
-internal object AdjustedRouteOptionsProvider {
+/**
+ * Default implementation of [RouteOptionsProvider]
+ */
+internal class MapboxRouteOptionsProvider(
+    private val logger: Logger
+) : RouteOptionsProvider {
 
-    private const val DEFAULT_REROUTE_BEARING_TOLERANCE = 90.0
+    private companion object {
+        const val DEFAULT_REROUTE_BEARING_TOLERANCE = 90.0
+    }
 
-    fun getRouteOptions(directionsSession: DirectionsSession, tripSession: TripSession, location: Location): RouteOptions? {
-        val routeOptions: RouteOptions = directionsSession.getRouteOptions() ?: return null
-        val routeProgress: RouteProgress = tripSession.getRouteProgress() ?: return null
+    /**
+     * Provide a new instance *RouteOptions* based on initial *RouteOptions*, *RouteProgress* and
+     * current *Location*
+     *
+     * Return *null* in case if cannot combine a new [RouteOptions] instance base on input.
+     */
+    override fun update(
+        routeOptions: RouteOptions?,
+        routeProgress: RouteProgress?,
+        location: Location?
```

Comment:
Can these be non-null to avoid checking for `null` 👇? Also after that, can `RouteOptions` be non-null too?

Reply:
should check that kind of checks(nonNull) out of invoker(MapboxNavigation)

Reply:
Yeah that's what I was thinking. IMO it adds clear expectations to clients implementing `RouteOptionsProvider` and also simplifies our `MapboxRouteOptionsProvider` 😅

Reply:
from client perspective to combine new route options any of this inputs might be null 😉 

But make FasterRouteController and MapboxNavigation more complex
>and also simplifies our MapboxRouteOptionsProvider 😅

Reply:
> from client perspective to combine new route options any of this inputs might be null 😉

Why? I don't fully understand. If `null` is not valid we can force it using non-nullable types.

> But make FasterRouteController and MapboxNavigation more complex

I don't agree. `FasterRouteController` and `MapboxNavigation` already handle those (they're already nullable) so it's a matter of checking in the call site but the contract is clear that those values need to be non-null to use it. I still don't see what's the use case for these to be nullable 🤔

Reply:
it's not fully true when we expose this interface for customers: client might combine new options and ignore any of inputs(routeOptions/routeProgress/location)

> I don't agree. FasterRouteController and MapboxNavigation already handle those (they're already nullable) so it's a matter of checking in the call site but the contract is clear that those values need to be non-null to use it. I still don't see what's the use case for these to be nullable 🤔

For the same reasons above

Reply:
> it's not fully true when we expose this interface for customers: client might combine new options and ignore any of inputs(routeOptions/routeProgress/location)

Could you provide an example? Not sure if that's a valid use case. I know that client can provide `RouteOptions` but what's the point of ignoring routeOptions / routeProgress / location altogether 🤔

Reply:
In case if we expose `RouteOptionsProvider` interface user might combine route options by his own. It means that he can ignore any inputs(routeOptions, routeProgress, or location) and provide new RouteOptions. So in that case inputs are null

Reply:
> Not sure if that's a valid use case. I know that client can provide `RouteOptions` but what's the point of ignoring routeOptions / routeProgress / location altogether 🤔